### PR TITLE
JDK-8277496 Remove duplication in c1 Block successor lists

### PR DESCRIPTION
**Pull request by LudwikJaniuk**
Remove `BlockBegin::_successors`, leaving `BlockEnd::_sux` as the SSOT for the successors of a block. Prior to this PR, these two lists were both tracking the same list of successors of the same block. This necessitated a lot of syncing and verification code.

With this PR, as long as a block has its end pointer assigned, its successors can always be reached by querying the `BlockEnd`.  `BlockEnd::_sux` becomes the single place where the list of successors is maintained. When modified, the successor list no longer needs to be synchronized in two places, reducing complexity and confusion. Asserts on the two lists corresponding no longer need to be made.

While being created in `GraphBuilder`, `BlockBegin`s don't have a `BlockEnd` assigned yet. To temporarily track block successors in this small interval, add a lookup structure `BlockListBuilder::_bci2block_successors`. 

This PR affects debug printing code. If the end pointer of a `BlockBegin `is NULL for some reason, then the successor list can no longer be printed (for obious reasons).

This PR introduces an additional check to IR::verify to check that `BlockBegin::_end` is not set to null.

This PR also performs some minor refactoring, polishing, inlining, and removing of dead code around the affected areas. 

The commit history has been polished to attempt to guide the reader through the changes.

hs-tier1 and hs-tier2 tests pass.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277496](https://bugs.openjdk.java.net/browse/JDK-8277496): Remove duplication in c1 Block successor lists

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6614/head:pull/6614` \
`$ git checkout pull/6614`

Update a local copy of the PR: \
`$ git checkout pull/6614` \
`$ git pull https://git.openjdk.java.net/jdk pull/6614/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6614`

View PR using the GUI difftool: \
`$ git pr show -t 6614`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6614.diff">https://git.openjdk.java.net/jdk/pull/6614.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/c1/c1_GraphBuilder.cpp [hunk 1]
```diff
@@ -3338,6 +3393,10 @@ GraphBuilder::GraphBuilder(Compilation* compilation, IRScope* scope)
 #ifndef PRODUCT
   if (PrintCompilation && Verbose) tty->print_cr("Created %d Instructions", _instruction_count);
 #endif
+
+  // JANIUK: If we iterate all the blocks in _blocks, some of them have end NULL.
```
Comment (neliasso) [inline src/hotspot/share/c1/c1_GraphBuilder.cpp:3397]: Left over comment?

**Review by neliasso — COMMENTED** [review]
A very nice clean up!

File: src/hotspot/share/c1/c1_GraphBuilder.cpp (hunk 1 above)
Comment (LudwikJaniuk) [reply src/hotspot/share/c1/c1_GraphBuilder.cpp]: Indeed! Removed it

File: src/hotspot/share/c1/c1_LIR.cpp [hunk 2]
```diff
@@ -1586,6 +1586,7 @@ static void print_block(BlockBegin* x) {
     }
   }
 
+  if (end != NULL)
```
Comment (neliasso) [inline src/hotspot/share/c1/c1_LIR.cpp:1589]: Missing braces on if statement

Comment (LudwikJaniuk) [reply src/hotspot/share/c1/c1_LIR.cpp]: Corrected it

**Review by neliasso — APPROVED** [review]
Looks good! Thanks for fixing!

**Review by vnkozlov — COMMENTED** [review]
@LudwikJaniuk  Please, rerun your testing. Running only `-b linux-x64` is not enough - you should not specify this option at all. Testing should be done on all platforms. Also `linux-x64` is product build which does not include asserts. So we may not catch possible issues.

**Review by vnkozlov — APPROVED** [review]
Good.
Testing results looks good too.

**Pull request closed without merging**